service/service_windows.go

```diff
@@ -0,0 +1,135 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build windows
+
+package service
+
+import (
+	"fmt"
+	"log"
+	"syscall"
+
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/eventlog"
+)
+
+type WindowsService struct {
+	app    *Application
+	params Parameters
+}
+
+func NewWindowsService(params Parameters) *WindowsService {
+	return &WindowsService{params: params}
+}
+
+func (s *WindowsService) Execute(args []string, requests <-chan svc.ChangeRequest, changes chan<- svc.Status) (ssec bool, errno uint32) {
+	if len(args) == 0 {
+		log.Fatal("expected first argument supplied to service.Execute to be the service name")
```

Comment:
Propagate error, here and everywhere else :)

Reply:
We can't really propogate errors in this file back to the main function from here as this function is invoked in a background goroutine via a syscall.

However, you did send me a on a path of improving how we handle errors here. Windows is not great at giving much information when services startup - you don't get any helpful error message from the startup command/dialog, & it takes a while to return any error. I tested the following:

1. **Log to event log (Event Viewer) directly:** Where possible this is probably the best option as the error will come up in Application event logs, and include whatever info you specify.
2. **Return a system errorcode:** This is useful to do where 1 is not possible or in addition to it. This will create an error representing the relevant error code in the System event logs.
3. **Use`log.Fatal`:** This is probably the worst option as you just get a generic error message in the System event logs with no details of the error.

So I've refactored this file quite a bit if you want to take another look